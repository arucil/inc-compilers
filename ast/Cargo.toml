[package]
name = "ast"
version = "0.1.0"
edition = "2018"

[dependencies]
pretty = "0.10.0"
support = { path = "../support" }

[dev-dependencies]
pretty_assertions = "0.7.2"
