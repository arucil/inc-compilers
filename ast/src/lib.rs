#![feature(never_type, box_syntax)]

use id_arena::{Arena, Id};
use pretty::*;
use std::fmt::{self, Debug, Formatter};
use std::iter;
use support::Range;

pub mod parser;

pub use parser::{parse, Result};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Program<VAR = String, TYPE = ()> {
  pub body: Vec<Exp<VAR, TYPE>>,
  pub types: Arena<CompType>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Exp<VAR = String, TYPE = ()> {
  pub kind: ExpKind<VAR, TYPE>,
  pub range: Range,
  pub ty: TYPE,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ExpKind<VAR, TYPE> {
  Int(i64),
  Prim {
    op: (Range, &'static str),
    args: Vec<Exp<VAR, TYPE>>,
  },
  Var(VAR),
  Str(String),
  Let {
    var: (Range, VAR),
    init: Box<Exp<VAR, TYPE>>,
    body: Box<Exp<VAR, TYPE>>,
  },
  Bool(bool),
  If {
    cond: Box<Exp<VAR, TYPE>>,
    conseq: Box<Exp<VAR, TYPE>>,
    alt: Box<Exp<VAR, TYPE>>,
  },
  Set {
    var: (Range, VAR),
    exp: Box<Exp<VAR, TYPE>>,
  },
  Get(VAR),
  Begin {
    seq: Vec<Exp<VAR, TYPE>>,
    last: Box<Exp<VAR, TYPE>>,
  },
  While {
    cond: Box<Exp<VAR, TYPE>>,
    body: Box<Exp<VAR, TYPE>>,
  },
  Void,
  Print(
    /// nonempty
    Vec<Exp<VAR, TYPE>>,
  ),
  NewLine,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Type {
  Int,
  Bool,
  Str,
  Comp(TypeId),
  Void,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TypeId(Id<CompType>);

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CompType {
  Vector(Vec<Type>),
}

#[derive(Clone, PartialEq, Eq, Hash)]
pub struct IdxVar {
  pub name: String,
  pub index: usize,
}

impl IdxVar {
  pub fn new<S: ToString>(name: S) -> Self {
    let name = name.to_string();
    if let Some(i) = name.find('.') {
      if let Ok(index) = name[i + 1..].parse::<usize>() {
        return Self {
          name: name[..i].to_owned(),
          index,
        };
      }
    }
    Self { name, index: 0 }
  }
}

impl Debug for IdxVar {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    write!(f, "{}.{}", self.name, self.index)
  }
}

impl<VAR: Debug, TYPE: Debug> Program<VAR, TYPE> {
  pub fn to_string_pretty(&self) -> String {
    let doc = self.to_doc();
    let mut buf = String::new();
    doc.render_fmt(30, &mut buf).unwrap();
    buf
  }

  fn to_doc(&self) -> RcDoc {
    RcDoc::intersperse(self.body.iter().map(|exp| exp.to_doc()), Doc::line())
  }
}

impl<VAR: Debug, TYPE: Debug> Exp<VAR, TYPE> {
  fn to_doc(&self) -> RcDoc {
    match &self.kind {
      ExpKind::Int(n) => RcDoc::text(format!("{}", n)),
      ExpKind::Str(s) => RcDoc::text(format!("{:?}", s)),
      ExpKind::Var(var) | ExpKind::Get(var) => {
        RcDoc::text(format!("{:?}", var))
      }
      ExpKind::Prim { op, args } => RcDoc::text("(")
        .append(
          RcDoc::intersperse(
            iter::once(RcDoc::text(op.1))
              .chain(args.iter().map(|arg| arg.to_doc())),
            Doc::line(),
          )
          .nest(1)
          .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Let { var, init, body } => RcDoc::text("(")
        .append(
          RcDoc::text("let")
            .append(Doc::line())
            .append(
              RcDoc::text("[")
                .append(
                  RcDoc::text(format!("{:?}", var.1))
                    .append(Doc::line())
                    .append(init.to_doc())
                    .nest(1)
                    .group(),
                )
                .append(RcDoc::text("]")),
            )
            .append(Doc::line())
            .append(body.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Bool(true) => RcDoc::text("#t"),
      ExpKind::Bool(false) => RcDoc::text("#f"),
      ExpKind::If { cond, conseq, alt } => RcDoc::text("(")
        .append(
          RcDoc::text("if")
            .append(Doc::line())
            .append(cond.to_doc())
            .group()
            .append(Doc::line())
            .append(conseq.to_doc().append(Doc::line()).append(alt.to_doc()))
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Set { var, exp } => RcDoc::text("(")
        .append(
          RcDoc::text("set!")
            .append(Doc::line())
            .append(RcDoc::text(format!("{:?}", var.1)))
            .append(Doc::line())
            .append(exp.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Begin { seq, last } => RcDoc::text("(")
        .append(
          RcDoc::text("begin").append(Doc::line()).append(
            RcDoc::intersperse(
              seq
                .iter()
                .chain(std::iter::once(&**last))
                .map(|exp| exp.to_doc()),
              Doc::line(),
            )
            .nest(1)
            .group(),
          ),
        )
        .append(RcDoc::text(")")),
      ExpKind::While { cond, body } => RcDoc::text("(")
        .append(
          RcDoc::text("while")
            .append(Doc::line())
            .append(cond.to_doc())
            .group()
            .append(Doc::line())
            .append(body.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Void => RcDoc::text("(void)"),
      ExpKind::Print(args) => RcDoc::text("(")
        .append(
          RcDoc::intersperse(
            iter::once(RcDoc::text("print"))
              .chain(args.iter().map(|arg| arg.to_doc())),
            Doc::line(),
          )
          .nest(1)
          .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::NewLine => RcDoc::text("(newline)"),
    }
  }
}
