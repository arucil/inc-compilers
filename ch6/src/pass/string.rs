use ast::{Error, Exp, ExpKind, FunDef, IdxVar, Program, Type};

pub fn expose_string_concat(
  prog: Program<IdxVar, Type>,
) -> Program<IdxVar, Type> {
  let mut state = State { tmp_counter: 0 };
  Program {
    fun_defs: prog
      .fun_defs
      .into_iter()
      .map(|(name, fun)| {
        (
          name,
          FunDef {
            body: state.exp_insert(fun.body),
            ..fun
          },
        )
      })
      .collect(),
    body: state.exp_insert(prog.body),
    ..prog
  }
}

struct State {
  tmp_counter: usize,
}

impl State {
  /// (string-append e1 e2)
  /// =>
  /// (let ([tmp1 e1]
  ///       [tmp2 e2]
  ///       [len1 (string-length tmp1)]
  ///       [result (alloc-string
  ///                 (+ len1 (string-length tmp2)))])
  ///   (copy-string! result 0 tmp1)
  ///   (copy-string! result len1 tmp2))
  fn exp_insert(&mut self, exp: Exp<IdxVar, Type>) -> Exp<IdxVar, Type> {
    match exp.kind {
      ExpKind::Int(_)
      | ExpKind::Var(_)
      | ExpKind::Str(_)
      | ExpKind::Bool(_)
      | ExpKind::Get(_)
      | ExpKind::Void
      | ExpKind::NewLine => exp,
      ExpKind::Prim {
        op: (_, "string-append"),
        mut args,
      } => {
        let range = exp.range;
        let arg2 = self.exp_insert(args.pop().unwrap());
        let arg1 = self.exp_insert(args.pop().unwrap());
        let tmp1 = self.new_tmp();
        let arg1_var = Exp {
          kind: ExpKind::Var(tmp1.clone()),
          range: exp.range,
          ty: Type::Int,
        };
        let tmp2 = self.new_tmp();
        let arg2_var = Exp {
          kind: ExpKind::Var(tmp2.clone()),
          range: exp.range,
          ty: Type::Int,
        };
        let len1 = self.new_tmp();
        let len1_var = Exp {
          kind: ExpKind::Var(len1.clone()),
          range: exp.range,
          ty: Type::Int,
        };
        let tmp3 = self.new_tmp();
        let result_var = Exp {
          kind: ExpKind::Var(tmp3.clone()),
          range: exp.range,
          ty: Type::Int,
        };
        let body = Exp {
          kind: ExpKind::Begin {
            seq: vec![
              Exp {
                kind: ExpKind::Prim {
                  op: (range, "copy-string!"),
                  args: vec![
                    result_var.clone(),
                    Exp {
                      kind: ExpKind::Int(0),
                      range,
                      ty: Type::Int,
                    },
                    arg1_var.clone(),
                  ],
                },
                range,
                ty: Type::Void,
              },
              Exp {
                kind: ExpKind::Prim {
                  op: (range, "copy-string!"),
                  args: vec![
                    result_var.clone(),
                    len1_var.clone(),
                    arg2_var.clone(),
                  ],
                },
                range,
                ty: Type::Void,
              },
            ],
            last: box result_var,
          },
          range,
          ty: Type::Str,
        };
        vec![
          (tmp1, box arg1),
          (tmp2, box arg2),
          (
            len1,
            box Exp {
              kind: ExpKind::Prim {
                op: (range, "string-length"),
                args: vec![arg1_var],
              },
              range,
              ty: Type::Int,
            },
          ),
          (
            tmp3,
            box Exp {
              kind: ExpKind::Prim {
                op: (exp.range, "alloc-string"),
                args: vec![Exp {
                  kind: ExpKind::Prim {
                    op: (exp.range, "+"),
                    args: vec![
                      len1_var,
                      Exp {
                        kind: ExpKind::Prim {
                          op: (range, "string-length"),
                          args: vec![arg2_var],
                        },
                        range,
                        ty: Type::Int,
                      },
                    ],
                  },
                  range: exp.range,
                  ty: Type::Int,
                }],
              },
              range,
              ty: Type::Str,
            },
          ),
        ]
        .into_iter()
        .rfold(body, |body, (var, init)| Exp {
          kind: ExpKind::Let {
            var: (range, var),
            init,
            body: box body,
          },
          range,
          ty: Type::Str,
        })
      }
      ExpKind::Prim { op, args } => Exp {
        kind: ExpKind::Prim {
          op,
          args: args.into_iter().map(|exp| self.exp_insert(exp)).collect(),
        },
        ..exp
      },
      ExpKind::Apply {
        fun,
        args,
        r#struct,
      } => Exp {
        kind: ExpKind::Apply {
          fun: box self.exp_insert(*fun),
          args: args.into_iter().map(|exp| self.exp_insert(exp)).collect(),
          r#struct,
        },
        ..exp
      },
      ExpKind::Let { var, init, body } => Exp {
        kind: ExpKind::Let {
          var,
          init: box self.exp_insert(*init),
          body: box self.exp_insert(*body),
        },
        ..exp
      },
      ExpKind::If { cond, conseq, alt } => Exp {
        kind: ExpKind::If {
          cond: box self.exp_insert(*cond),
          conseq: box self.exp_insert(*conseq),
          alt: box self.exp_insert(*alt),
        },
        ..exp
      },
      ExpKind::Set { var, exp: exp1 } => Exp {
        kind: ExpKind::Set {
          var,
          exp: box self.exp_insert(*exp1),
        },
        ..exp
      },
      ExpKind::Begin { seq, last } => Exp {
        kind: ExpKind::Begin {
          seq: seq.into_iter().map(|exp| self.exp_insert(exp)).collect(),
          last: box self.exp_insert(*last),
        },
        ..exp
      },
      ExpKind::While { cond, body } => Exp {
        kind: ExpKind::While {
          cond: box self.exp_insert(*cond),
          body: box self.exp_insert(*body),
        },
        ..exp
      },
      ExpKind::Print(args) => Exp {
        kind: ExpKind::Print(
          args.into_iter().map(|exp| self.exp_insert(exp)).collect(),
        ),
        ..exp
      },
      ExpKind::Error(Error::Length(arg)) => Exp {
        kind: ExpKind::Error(Error::Length(box self.exp_insert(*arg))),
        ..exp
      },
      ExpKind::Error(Error::OutOfBounds { index, len }) => Exp {
        kind: ExpKind::Error(Error::OutOfBounds {
          index: box self.exp_insert(*index),
          len: box self.exp_insert(*len),
        }),
        ..exp
      },
      ExpKind::Error(Error::DivByZero) => exp,
      ExpKind::FunRef { .. } => exp,
    }
  }

  fn new_tmp(&mut self) -> IdxVar {
    self.tmp_counter += 1;
    IdxVar {
      name: "(str-tmp)".to_owned(),
      index: self.tmp_counter,
    }
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use ch4::pass::uniquify;
  use ch5::pass::typecheck::typecheck;
  use insta::assert_snapshot;

  #[test]
  fn string_append() {
    let prog = parse(
      r#"
(let ([x "\n"])
  (print (string-append "abc" (string-append "+" x))))
      "#,
    )
    .unwrap();
    let prog = typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let result = expose_string_concat(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
