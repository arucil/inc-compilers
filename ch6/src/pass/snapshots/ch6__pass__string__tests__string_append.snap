---
source: ch6/src/pass/string.rs
expression: result.to_string_pretty()
---
(let
 [x.0 "\n"]
 (print
  (let
   [(str-tmp).5 "abc"]
   (let
    [(str-tmp).6
     (let
      [(str-tmp).1 "+"]
      (let
       [(str-tmp).2 x.0]
       (let
        [(str-tmp).3 (string-length (str-tmp).1)]
        (let
         [(str-tmp).4
          (alloc-string
           (+
            (str-tmp).3
            (string-length (str-tmp).2)))]
         (begin
          (copy-string! (str-tmp).4 0 (str-tmp).1)
          (copy-string!
           (str-tmp).4
           (str-tmp).3
           (str-tmp).2)
          (str-tmp).4)))))]
    (let
     [(str-tmp).7 (string-length (str-tmp).5)]
     (let
      [(str-tmp).8
       (alloc-string
        (+
         (str-tmp).7
         (string-length (str-tmp).6)))]
      (begin
       (copy-string! (str-tmp).8 0 (str-tmp).5)
       (copy-string!
        (str-tmp).8
        (str-tmp).7
        (str-tmp).6)
       (str-tmp).8)))))))
